cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/WorkflowStatsSLAHttpHandler.java

```diff
@@ -89,4 +108,133 @@ public void workflowStats(HttpRequest request, HttpResponder responder,
+                      @PathParam("run-id") String runId,
+                      @QueryParam("other-run-id") String otherRunId) throws Exception {
+    Id.Workflow id = Id.Workflow.from(Id.Namespace.from(namespaceId), appId, workflowId);
+    WorkflowStatistics.DetailedStatistics detailedStatistics = getDetailedRecord(id, runId);
+    WorkflowStatistics.DetailedStatistics otherDetailedStatistics = getDetailedRecord(id, otherRunId);
+    if (detailedStatistics == null) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, "The run-id you provided was not correct.");
+      return;
+    }
+    if (otherDetailedStatistics == null) {
+      responder.sendString(HttpResponseStatus.BAD_REQUEST, "The other run-id you provided was not correct.");
+      return;
+    }
+
+    Map<String, WorkflowStatistics.DetailedStatistics> detailedStatisticsMap = new HashMap<>();
+    detailedStatisticsMap.put(runId, detailedStatistics);
+    detailedStatisticsMap.put(otherRunId, otherDetailedStatistics);
+
+    Map<String, Map<String, Map<String, Long>>> formattedStatisticsMap = convert(detailedStatisticsMap);
+    responder.sendJson(HttpResponseStatus.OK, formattedStatisticsMap);
+  }
+
+  private Map<String, Map<String, Map<String, Long>>> convert(Map<String, WorkflowStatistics.DetailedStatistics>
+                                                                workflowDetailedStatisticsMap) {
+    Map<String, Map<String, Map<String, Long>>> formattedStatisticsMap = new HashMap<>();
+    for (Map.Entry<String, WorkflowStatistics.DetailedStatistics> entry : workflowDetailedStatisticsMap.entrySet()) {
+      Map<String, Map<String, Long>> internalMap = entry.getValue().getProgramToStatistics();
+      for (Map.Entry<String, Map<String, Long>> internalEntry : internalMap.entrySet()) {
+        Map<String, Map<String, Long>> newMap;
+        if ((newMap = formattedStatisticsMap.get(internalEntry.getKey())) == null) {
+          newMap = new HashMap<>();
+          formattedStatisticsMap.put(internalEntry.getKey(), newMap);
+        }
+        newMap.put(entry.getKey(), internalEntry.getValue());
+      }
+    }
+    return formattedStatisticsMap;
+  }
+
+  @Nullable
+  public WorkflowStatistics.DetailedStatistics getDetailedRecord(Id.Workflow workflowId,
+                                                                 String runId) throws Exception {
+    WorkflowDataset.WorkflowRunRecord workflowRunRecord = store.getWorkflowRun(workflowId, runId);
+    if (workflowRunRecord == null) {
+      return null;
+    }
+    List<WorkflowDataset.ProgramRun> programRuns = workflowRunRecord.getProgramRuns();
+    Map<String, Map<String, Long>> details = new HashMap<>();
+    for (WorkflowDataset.ProgramRun actionRun : programRuns) {
+      if (actionRun.getProgramType() == ProgramType.MAPREDUCE) {
+        details.put(actionRun.getName(),
+                    getMapreduceDetails(Id.Program.from(workflowId.getNamespaceId(), workflowId.getApplicationId(),
```

Comment:
can extract ``Id.Program.from(workflowId.getNamespaceId(), workflowId.getApplicationId(), actionRun.getProgramType(), actionRun.getName()`` into a var so you don't have to repeat it for mapreduce and spark